browser/js/dashboard/gridbox/WidgetSettingsFactory.js

```diff
@@ -0,0 +1,36 @@
+app.factory('WidgetSettingsFactory', function($interval, DashboardFactory){
+  var factory = {};
+
+  factory.newSetKeys = function(dataSource){
+    return DashboardFactory.getDataSource(dataSource)
+      .then(function(data){
+        var realData = factory.findDataToGraph(data);
+        if (!realData) {
```

Comment:
Could the `.findDataToGraph` method return the original data if it cannot otherwise find "realData"? That way you wouldn't have to check here, and so using the utility method would be more straightforward.